Debug assertion failure: QuadraticSolver.hh:110: `std::fabs(a) >= min_a()`

#### Failure:
```json
    "failure": {
     "stderr": [
      "/home/alund/celeritas_project/celeritas/app/demo-loop/LDemoIO.cc:236: warning: Geometry contains surfaces that are incompatible with the current ORANGE simple safety algorithm: multiple scattering may result in arbitrarily small steps",
      "terminate called after throwing an instance of 'celeritas::RuntimeError'",
      "  what():  /home/alund/celeritas_project/celeritas/src/corecel/data/DeviceAllocation.cc:61:",
      "celeritas: CUDA error: cudaFree(ptr) failed:",
      "[PID=36248, TID=-1][ 7/15]> /soft/compilers/gcc/11.1.0/x86_64-suse-linux/lib64/libgcc_s.so.1(_Unwind_RaiseException+0x311) [0x7f76020f24d1]"
     ],   
     "stdout": [
      "/home/alund/celeritas_project/celeritas/src/orange/surf/detail/../detail/QuadraticSolver.hh:110:",
      "celeritas: internal assertion failed: std::fabs(a) >= min_a()"
     ]    
    } 
```
#### Regression problem:
simple-cms+field-orange-gpu

#### Diagnosis:
This shows up when calculating the intersections with a z-aligned cylinder for a particle with z-direction very close to 1: `a = 1 - ipow<2>(dir[z])` = 1.18371978885499152e-12 is less than  the quadratic solver `min_a` = 1e-10.

position = (9.89771847341366993e+01, 1.06416436785971431e+01, 4.46914463342164936e+01)
direction = (4.09517743700767399e-07, 1.00812588415826643e-06, 9.99999999999408140e-01)

--- a/src/orange/surf/CylCentered.hh
+++ b/src/orange/surf/CylCentered.hh
@@ -160,7 +160,7 @@ CylCentered<T>::calc_intersections(Real3 const& pos,
     // 1 - \omega \dot e
     const real_type a = 1 - ipow<2>(dir[t_index()]);
 
-    if (a != 0)
+    if (a >= detail::QuadraticSolver::min_a())
     {
         const real_type u = pos[u_index()];
         const real_type v = pos[v_index()];

--- a/test/orange/surf/CylCentered.test.cc
+++ b/test/orange/surf/CylCentered.test.cc
@@ -380,7 +380,6 @@ TEST(TestCCylZ, multi_along_intersect)
     CCylZ cyl(30.0);
 
     VecReal all_first_distances;
-    VecReal all_errors;
 
     for (real_type sqrt_eps : {1e-1, 1e-2, 1e-4, 1e-5, 1e-6, 1e-7})
     {
@@ -405,24 +404,26 @@ TEST(TestCCylZ, multi_along_intersect)
                 if (d == no_intersection())
                     continue;
 
+                // Move the solved distance and make sure we're close to the
+                // actual cylinder surface
                 axpy(d, dir, &pos);
                 real_type boundary_error = std::hypot(pos[0], pos[1])
                                            - real_type(30);
-                all_errors.push_back(
-                    boundary_error
-                    / (std::numeric_limits<real_type>::epsilon() * d));
 
-                EXPECT_LT(std::fabs(boundary_error), real_type(1))
-                    << "huge error " << boundary_error << " from " << orig_pos
-                    << " along " << dir << " with calculated distance " << d;
+                // Calculate relative to the distance traveled and floating
+                // point precision
+                real_type rel_error_ulp
+                    = boundary_error
+                      / (std::numeric_limits<real_type>::epsilon() * d);
+                EXPECT_LT(std::fabs(rel_error_ulp), 2000)
+                    << "large error " << boundary_error << " ("
+                    << rel_error_ulp << ") ULP from " << orig_pos << " along "
+                    << dir << " with calculated distance " << d;
             }
         }
     }
 
     constexpr real_type inf = no_intersection();
-
-    // For long distances we can overshoot or undershoot due to numerical
-    // error.
 
     // clang-format off
     static const double expected_all_first_distances[] = {inf, 2191.4760245467,
@@ -437,17 +438,9 @@ TEST(TestCCylZ, multi_along_intersect)
         768114.54300982, inf, inf, 9.99984331429e-06, 7.7459254135055, inf,
         inf, inf, inf, inf, inf, inf, inf, inf, inf, inf, inf, inf, inf, inf,
         inf, inf, inf, inf, inf, inf, inf, inf, inf};
-    static const double expected_all_errors[] = {52.063540153765,
-        -0.46230400035493, 3.3838561098844, 2.3998800089991, 0, 1.243610958517,
-        -15.999200059996, 0, 0, -645.19304347809, 25.058574704815,
-        50.290552485276, -29.443199854709, 1.7886888728781, 18.246365467577, 0,
-        -1599.8426904211, 0, 561.93151553504, -623.35388309947,
-        40.334088238136, -226.15209393773, 0, 0, -1706.6981779472,
-        -1316.4685033787, -122.50084953256, -477.61345405917, 0, 0};
     // clang-format on
 
     EXPECT_VEC_NEAR(expected_all_first_distances, all_first_distances, 1e-5);
-    EXPECT_VEC_NEAR(expected_all_errors, all_errors, 1e-2);
 }
 
 //---------------------------------------------------------------------------//

--- a/test/orange/surf/CylCentered.test.cc
+++ b/test/orange/surf/CylCentered.test.cc
@@ -11,7 +11,9 @@
 #include <cmath>
 #include <vector>
 
+#include "corecel/cont/ArrayIO.hh"
 #include "corecel/math/Algorithms.hh"
+#include "corecel/math/ArrayUtils.hh"
 
 #include "celeritas_test.hh"
 
@@ -23,6 +25,19 @@ namespace test
 
 using Intersections = CCylX::Intersections;
 using VecReal = std::vector<real_type>;
+
+real_type min_intersection(Intersections const& i)
+{
+    if (i[0] == 0 && i[1] == 0)
+        return no_intersection();
+    else if (i[0] == 0)
+        return i[1];
+    else if (i[1] == 0)
+        return i[0];
+    else if (i[0] < i[1])
+        return i[0];
+    return i[1];
+}
 
 //---------------------------------------------------------------------------//
 // TESTS
@@ -98,6 +113,23 @@ TEST(TestCCylX, intersect)
 
     EXPECT_EQ(no_intersection(), distances[0]);
     EXPECT_EQ(no_intersection(), distances[1]);
+
+    // From inside, exactly along the axis
+    distances[0] = distances[1] = -1;
+    distances = cyl.calc_intersections(
+        Real3{0.123, 0.345, 0.456}, Real3{1, 0, 0}, SurfaceState::off);
+    EXPECT_EQ(no_intersection(), distances[0]);
+    EXPECT_EQ(no_intersection(), distances[1]);
+
+    // From inside, nearly along the axis
+    distances[0] = distances[1] = -1;
+    distances = cyl.calc_intersections(Real3{0.123, 0.345, 0.456},
+                                       Real3{9.99999999999408140e-01,
+                                             4.09517743700767399e-07,
+                                             1.00812588415826643e-06},
+                                       SurfaceState::off);
+    EXPECT_EQ(no_intersection(), distances[0]);
+    EXPECT_SOFT_EQ(no_intersection(), distances[1]);
 }
 
 TEST(TestCCylX, intersect_from_surface)
@@ -285,7 +317,7 @@ TEST(TestCCylZ, calc_intersections_on_surface)
 
 // Fused multiply-add on some CPUs in opt mode can cause the results of
 // nearly-tangent cylinder checking to change.
-TEST(TestCCylZ, multi_intersect)
+TEST(TestCCylZ, multi_tangent_intersect)
 {
     constexpr int Y = static_cast<int>(Axis::y);
 
@@ -301,68 +333,93 @@ TEST(TestCCylZ, multi_intersect)
             Real3 pos{x, -10.0001 * v, 0};
             Real3 dir{0, v, 0};
 
-            Intersections d;
+            real_type d;
 
             // Transport to inside of cylinder
-            d = cyl.calc_intersections(pos, dir, SurfaceState::off);
-            ASSERT_NE(no_intersection(), d[0]);
-            all_first_distances.push_back(d[0]);
-            pos[Y] += d[0] * dir[Y];
+            d = min_intersection(
+                cyl.calc_intersections(pos, dir, SurfaceState::off));
+            ASSERT_NE(no_intersection(), d);
+            all_first_distances.push_back(d);
+            pos[Y] += d * dir[Y];
             all_y.push_back(pos[Y]);
 
             // Transport to other side of cylinder
-            d = cyl.calc_intersections(pos, dir, SurfaceState::on);
-            all_distances.push_back(d[0]);
-            ASSERT_NE(no_intersection(), d[0]);
-            pos[Y] += d[0] * dir[Y];
+            d = min_intersection(
+                cyl.calc_intersections(pos, dir, SurfaceState::on));
+            all_distances.push_back(d);
+            if (d == no_intersection())
+                continue;
+
+            pos[Y] += d * dir[Y];
 
             // We're done
-            d = cyl.calc_intersections(pos, dir, SurfaceState::on);
-            EXPECT_EQ(no_intersection(), d[0]);
+            d = min_intersection(
+                cyl.calc_intersections(pos, dir, SurfaceState::on));
+            EXPECT_EQ(no_intersection(), d);
         }
     }
 
-    const real_type expected_all_first_distances[] = {9.99869,
-                                                      9.99869,
-                                                      0.0001,
-                                                      0.0001,
-                                                      0.0001,
-                                                      0.0001,
-                                                      0.0001,
-                                                      0.0001,
-                                                      0.0001,
-                                                      0.0001,
-                                                      9.99869,
-                                                      9.99869};
+    // clang-format off
+    const real_type expected_all_first_distances[] = {9.99869, 9.99869, 0.0001,
+        0.0001, 0.0001, 0.0001, 0.0001, 0.0001, 0.0001, 0.0001, 9.99869,
+        9.99869};
     EXPECT_VEC_NEAR(expected_all_first_distances, all_first_distances, 1e-5);
 
-    const real_type expected_all_y[] = {0.00141421,
-                                        -0.00141421,
-                                        10,
-                                        -10,
-                                        10,
-                                        -10,
-                                        10,
-                                        -10,
-                                        10,
-                                        -10,
-                                        0.00141421,
-                                        -0.00141421};
+    const real_type expected_all_y[] = {0.00141421, -0.00141421, 10, -10, 10,
+        -10, 10, -10, 10, -10, 0.00141421, -0.00141421};
     EXPECT_VEC_NEAR(expected_all_y, all_y, 1e-5);
 
-    const real_type expected_all_distances[] = {0.00282843,
-                                                0.00282843,
-                                                20,
-                                                20,
-                                                20,
-                                                20,
-                                                20,
-                                                20,
-                                                20,
-                                                20,
-                                                0.00282843,
-                                                0.00282843};
+    const real_type expected_all_distances[] = {0.00282843, 0.00282843, 20,
+        20, 20, 20, 20, 20, 20, 20, 0.00282843, 0.00282843};
     EXPECT_VEC_NEAR(expected_all_distances, all_distances, 1e-5);
+    // clang-format on
+}
+
+TEST(TestCCylZ, multi_along_intersect)
+{
+    CCylZ cyl(30.0);
+
+    VecReal all_first_distances;
+    VecReal all_errors;
+
+    for (real_type sqrt_eps : {1e-1, 1e-2, 1e-4, 1e-5, 1e-6, 1e-7})
+    {
+        real_type eps = ipow<2>(sqrt_eps);
+        for (Real3 const& orig_pos :
+             {Real3{7, 8, 9}, Real3{29, 0, 0}, Real3{30 - eps, 0, 0}})
+        {
+            for (Real3 dir : {
+                     Real3{0, 0, 1},
+                     Real3{eps, 0, 1},
+                     Real3{sqrt_eps, sqrt_eps, 1},
+                     Real3{0, sqrt_eps, 1},
+                 })
+            {
+                Real3 pos = orig_pos;
+                normalize_direction(&dir);
+
+                real_type d;
+
+                // Transport to inside of cylinder
+                d = min_intersection(
+                    cyl.calc_intersections(pos, dir, SurfaceState::off));
+                all_first_distances.push_back(d);
+                if (d == no_intersection())
+                    continue;
+
+                axpy(d, dir, &pos);
+
+                real_type boundary_error = std::hypot(pos[0], pos[1])
+                                           - real_type(30);
+                all_errors.push_back(
+                    boundary_error
+                    / (std::numeric_limits<real_type>::epsilon() * d));
+            }
+        }
+    }
+
+    PRINT_EXPECTED(all_first_distances);
+    PRINT_EXPECTED(all_errors);
 }
 
 //---------------------------------------------------------------------------//

--- a/test/orange/surf/CylCentered.test.cc
+++ b/test/orange/surf/CylCentered.test.cc
@@ -398,28 +398,56 @@ TEST(TestCCylZ, multi_along_intersect)
                 Real3 pos = orig_pos;
                 normalize_direction(&dir);
 
-                real_type d;
-
                 // Transport to inside of cylinder
-                d = min_intersection(
+                real_type d = min_intersection(
                     cyl.calc_intersections(pos, dir, SurfaceState::off));
                 all_first_distances.push_back(d);
                 if (d == no_intersection())
                     continue;
 
                 axpy(d, dir, &pos);
-
                 real_type boundary_error = std::hypot(pos[0], pos[1])
                                            - real_type(30);
                 all_errors.push_back(
                     boundary_error
                     / (std::numeric_limits<real_type>::epsilon() * d));
+
+                EXPECT_LT(std::fabs(boundary_error), real_type(1))
+                    << "huge error " << boundary_error << " from " << orig_pos
+                    << " along " << dir << " with calculated distance " << d;
             }
         }
     }
 
-    PRINT_EXPECTED(all_first_distances);
-    PRINT_EXPECTED(all_errors);
+    constexpr real_type inf = no_intersection();
+
+    // For long distances we can overshoot or undershoot due to numerical
+    // error.
+
+    // clang-format off
+    static const double expected_all_first_distances[] = {inf, 2191.4760245467,
+        138.43704564716, 212.77500479316, inf, 100.00499987501,
+        9.9365248561264, 77.194559393782, inf, 1.00004999875, 0.10097822248343,
+        7.7839514386978, inf, 219136.64666597, 1370.868072293, 2117.2962860405,
+        inf, 10000.000049346, 98.396094853996, 768.15297955672, inf,
+        1.0000000050059, 0.010000983281543, 7.7463475257587, inf, inf,
+        137073.10209619, 211719.04367353, inf, inf, 9838.6257712601,
+        76811.457712096, inf, inf, 0.00010000000474975, 7.7459668529435, inf,
+        inf, 1370730.9694406, 2117190.3656234, inf, inf, 98386.256470787,
+        768114.54300982, inf, inf, 9.99984331429e-06, 7.7459254135055, inf,
+        inf, inf, inf, inf, inf, inf, inf, inf, inf, inf, inf, inf, inf, inf,
+        inf, inf, inf, inf, inf, inf, inf, inf, inf};
+    static const double expected_all_errors[] = {52.063540153765,
+        -0.46230400035493, 3.3838561098844, 2.3998800089991, 0, 1.243610958517,
+        -15.999200059996, 0, 0, -645.19304347809, 25.058574704815,
+        50.290552485276, -29.443199854709, 1.7886888728781, 18.246365467577, 0,
+        -1599.8426904211, 0, 561.93151553504, -623.35388309947,
+        40.334088238136, -226.15209393773, 0, 0, -1706.6981779472,
+        -1316.4685033787, -122.50084953256, -477.61345405917, 0, 0};
+    // clang-format on
+
+    EXPECT_VEC_NEAR(expected_all_first_distances, all_first_distances, 1e-5);
+    EXPECT_VEC_NEAR(expected_all_errors, all_errors, 1e-2);
 }
 
 //---------------------------------------------------------------------------//

